### ISSUE TITLE
OSD upgrade failure from 1.0 to 1.1 with "managedFields" on

### ISSUE THREAD

**Issue by yanchicago** [rook/rook#4364]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
During upgrade from 1.0.2 to 1.1.7 upgrade, the mon/mgr deploy were upgraded successfully but osd deployment failed. Seems that rook upgrade doesn't handle "managed field" included in the metadata. 

**Expected behavior:**
OSD upgrade should succeed. 

**How to reproduce it (minimal and precise):**

1). Deploy rook 1.0.2 cluster
2). Upgrade operator image from 1.0.2 to 1.1.7
3). The osd deploy failed to be upgraded with below failure shown in the operator log. Note: The mon/mgr deploy succeeded with the upgrade. 
```
2019-11-22 15:54:24.340219 I | op-osd: starting 1 osd daemons on node olive-1906-e-s-01
2019-11-22 15:54:24.369473 W | op-osd: failed to create osd deployment for node olive-1906-e-s-01, osd {1 /var/lib/rook/osd1 /var/lib/rook/osd1/rook-ceph.config ceph /var/lib/rook/osd1/keyring 15dcf396-b263-4145-90d7-4a1fe7ea188c  false false  true  false}: failed to update object (Create for apps/v1, Kind=Deployment) managed fields: failed to create typed new object: .spec.template.spec.containers[name="osd"].env: duplicate entries for key [name="NODE_NAME"]
2019-11-22 15:54:24.377129 I | op-osd: osd orchestration status for node olive-1906-e-s-02 is completed
2019-11-22 15:54:24.377148 I | op-osd: starting 1 osd daemons on node olive-1906-e-s-02
2019-11-22 15:54:24.381468 W | op-osd: failed to create osd deployment for node olive-1906-e-s-02, osd {0 /var/lib/rook/osd0 /var/lib/rook/osd0/rook-ceph.config ceph /var/lib/rook/osd0/keyring bd8cdeb6-4419-422f-a5ec-2a49a35d1db4  false false  true  false}: failed to update object (Create for apps/v1, Kind=Deployment) managed fields: failed to create typed new object: .spec.template.spec.containers[name="osd"].env: duplicate entries for key [name="NODE_NAME"]
```
```
# kubectl -n rook-ceph get deployments -l rook_cluster=rook-ceph -o jsonpath='{range .items[*]}{.metadata.name}{"\trook-version="}{.metadata.labels.rook-version}{"\n"}{end}'
rook-ceph-mgr-a rook-version=v1.1.7
rook-ceph-mon-a rook-version=v1.1.7
rook-ceph-mon-b rook-version=v1.1.7
rook-ceph-mon-c rook-version=v1.1.7
rook-ceph-osd-0 rook-version=v1.0.2
rook-ceph-osd-1 rook-version=v1.0.2
rook-ceph-osd-2 rook-version=v1.0.2
rook-ceph-osd-3 rook-version=v1.0.2
rook-ceph-osd-4 rook-version=v1.0.2
rook-ceph-rgw-rook-ceph-store   rook-version=v1.0.2
```

4). The failed osd has "managed field" included in the metadata.  In the different k8s cluster, the osds without "managed field" would succeed. 
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
  creationTimestamp: "2019-11-22T07:15:38Z"
  generation: 1
  labels:
    app: rook-ceph-osd
    ceph-osd-id: "0"
    ceph-version: 14.2.1
    rook-version: v1.0.2
    rook_cluster: rook-ceph
  managedFields:
  - apiVersion: apps/v1
    fields:
      f:metadata:
        f:annotations:
          .: null
          f:deployment.kubernetes.io/revision: null
```

5). The first "NODE_NAME" is in "managedFields". And the 2nd is regular one. 
```
#]# knc get pod rook-ceph-osd-0-57b64b6587-4g9lr -o json | grep -n "\"NODE_NAME"
66:                                    "k:{\"name\":\"NODE_NAME\"}": {
755:                        "name": "NODE_NAME",
```
```
  "managedFields": [
            {
                "apiVersion": "v1",
                "fields": {
                    "f:metadata": {
                        "f:annotations": {
                            ".": null,
                            "f:kubernetes.io/psp": null
                        },
                        "f:generateName": null,
                        "f:labels": {
                            ".": null,
                            "f:app": null,
                            "f:ceph-osd-id": null,
                            "f:pod-template-hash": null,
                            "f:rook_cluster": null
                        },
                        "f:ownerReferences": {
                            ".": null,
                            "k:{\"uid\":\"e2673ba6-0cf7-11ea-b07c-fa163e064919\"}": {
                                ".": null,
                                "f:apiVersion": null,
                                "f:blockOwnerDeletion": null,
                                "f:controller": null,
                                "f:kind": null,
                                "f:name": null,
                                "f:uid": null
                            }
                        }
                    },
                    "f:spec": {
                        "f:affinity": {
                            ".": null,
                            "f:nodeAffinity": {
                                ".": null,
                                "f:requiredDuringSchedulingIgnoredDuringExecution": null
                            }
                        },
                        "f:containers": {
                            "k:{\"name\":\"osd\"}": {
                                ".": null,
                                "f:args": null,
                                "f:command": null,
                                "f:env": {
                                    ".": null,
                                    "k:{\"name\":\"CONTAINER_IMAGE\"}": {
                                        ".": null,
                                        "f:name": null,
                                        "f:value": null
                                    },
                                    "k:{\"name\":\"NODE_NAME\"}": {
                                        ".": null,
                                        "f:name": null,
                                        "f:valueFrom": {
```

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release): centos 7.6
* Kernel (e.g. `uname -a`): 
* Cloud provider or hardware configuration: openstack
* Rook version (use `rook version` inside of a Rook Pod): rook1.1.7
* Storage backend version (e.g. for ceph do `ceph -v`): 14.2.1
* Kubernetes version (use `kubectl version`): 14.3
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Tectonic
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

**Comment by travisn** [comment]
@rohan47 Could you take a look at this to get a repro?

**Comment by rohan47** [comment]
@travisn Looking into it

**Comment by rohan47** [comment]
@yanchicago can you please add more info on how to repro this?
In my case I update operator image using `kubectl -n rook-ceph set image deploy/rook-ceph-operator rook-ceph-operator=rook/ceph:v1.1.7` and only the operator, discover and agent pods gets updated.

**Comment by yanchicago** [comment]
When ‘ServerSideApply=true’ feature-gates turned on in k8s, "managedFields" will show.

**Comment by rohan47** [comment]
I was not able to repro this.
The steps that I followed are
1. checked release-1.0 branch changed the operator image to 1.0.2 from 1.0.6 in operator.yaml and deployed the rook cluster.
2. checked release-1.1 branch and followed instructions from the [upgrade doc](https://rook.io/docs/rook/v1.1/ceph-upgrade.html) with a change in **Update the Rook Operator** step `kubectl -n $ROOK_SYSTEM_NAMESPACE set image deploy/rook-ceph-operator rook-ceph-operator=rook/ceph:v1.1.7`

```
$ kubectl -n rook-ceph get deployments -l rook_cluster=rook-ceph -o jsonpath={range .items[*]}{.metadata.name}{"...  localhost.localdomain: Tue Dec  3 12:08:05 2019

rook-ceph-mgr-a         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-mon-a         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-mon-b         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-mon-c         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-0         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-1         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-2         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-3         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-4         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-5         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-6         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-7         req/upd/avl: 1/1/1      rook-version=v1.1.7
rook-ceph-osd-8         req/upd/avl: 1/1/1      rook-version=v1.1.7
```

**Comment by travisn** [comment]
@rohan47 Did you enable the `ServerSideApply` feature gate as mentioned above?

**Comment by rohan47** [comment]
to enable the `ServerSideApply` feature gate I have to pass it to `kubeadm init` as `--feature-gates=ServerSideApply=true` right? I guess it not getting enabled.

**Comment by yanchicago** [comment]
This feature gate will be true by default in kubernetes 1.16. 
< ServerSideApply | true | Beta | 1.16

In my setup, I  updated  "/etc/kubernetes/manifests/kube-apiserver.yml" to add feature-gates with ServerSideApply=true. Then restart kubelet which will restart apiserver.

**Comment by rohan47** [comment]
I am using kubernetes 1.16, so it should be true by default

**Comment by yanchicago** [comment]
Do you see "managedFields" for osd pod? It should apply to any k8s resources:  node, deploy? If not, you may enable it to make sure "managedFields" appear. 
I'm testing with k8s 14.3 release.

**Comment by rohan47** [comment]
I am not seeing the managed fields even after enabling ServerSideApply
```ps aux | grep kubelet | grep feature
root      2091  7.5 19.1 550700 359480 ?       Ssl  05:28   0:21 kube-apiserver --advertise-address=192.168.73.100 --allow-privileged=true --authorization-mode=Node,RBAC --client-ca-file=/etc/kubernetes/pki/ca.crt --enable-admission-plugins=NodeRestriction --enable-bootstrap-token-auth=true --etcd-cafile=/etc/kubernetes/pki/etcd/ca.crt --etcd-certfile=/etc/kubernetes/pki/apiserver-etcd-client.crt --etcd-keyfile=/etc/kubernetes/pki/apiserver-etcd-client.key --etcd-servers=https://127.0.0.1:2379 --insecure-port=0 --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key --requestheader-allowed-names=front-proxy-client --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-group-headers=X-Remote-Group --requestheader-username-headers=X-Remote-User --secure-port=6443 --service-account-key-file=/etc/kubernetes/pki/sa.pub --service-cluster-ip-range=10.96.0.0/12 --tls-cert-file=/etc/kubernetes/pki/apiserver.crt --tls-private-key-file=/etc/kubernetes/pki/apiserver.key 
--feature-gates=ServerSideApply=true

```

**Comment by yanchicago** [comment]
In 14.3, with  "AllAlpha=true" , the "managed fields" will show. And it's disabled by "AllAlpha=true, ServerSideApply=false".

**Comment by rohan47** [comment]
Was able to repro this in k8s release 1.14.9

**Comment by yanchicago** [comment]
Please keep it open as the feature will be on in the future.

**Comment by yanchicago** [comment]
The "ServerSideApply" will be the default in k8s 1.18 release and not recommended to be turned off. 
Could you please support the config?

**Comment by travisn** [comment]
I believe this was fixed by #6272, the fix will be included in v1.4.5. Please reopen if it's still an issue after that.
